telemetry/Tracks/class-tracks-event.php

```diff
@@ -0,0 +1,275 @@
+
+	/**
+	 * Property name regex.
+	 */
+	protected const PROPERTY_NAME_REGEX = '/^[a-z_][a-z0-9_]*$/';
+
+	/**
+	 * Variable containing the event's data or an error if one was encountered
+	 * during the event's creation.
+	 *
+	 * @var stdClass|WP_Error
+	 */
+	protected $data = null;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param string                            $event_name The event's name.
+	 * @param array<string, mixed>|array<empty> $event_properties Any properties included in the event.
+	 */
+	public function __construct( string $event_name, array $event_properties ) {
+		$event_data        = static::process_properties( $event_name, $event_properties );
+		$validation_result = static::get_event_validation_result( $event_data );
+
+		$this->data = $validation_result ?? $event_data;
+	}
+
+	/**
+	 * Returns the event's data.
+	 *
+	 * @return stdClass|WP_Error Event object if the event was created successfully, WP_Error otherwise.
+	 */
+	public function get_data() {
+		return $this->data;
+	}
+
+	/**
+	 * Returns whether the event can be recorded.
+	 *
+	 * @return bool|WP_Error True if the event is recordable.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if the event is generating an error.
+	 */
+	public function is_recordable() {
+		if ( ! Telemetry_System::is_wpadmin_telemetry_allowed() ) {
+			return false;
+		}
+
+		// Don't record events during unit tests and CI runs.
+		if ( 'wptests_capabilities' === wp_get_current_user()->cap_key ) {
+			return false;
+		}
+
+		if ( is_wp_error( $this->data ) ) {
+			return $this->data;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Processes the event's properties to get them ready for validation.
+	 *
+	 * @param string $event_name The event's name.
+	 * @param array<string, mixed>|array<empty> $event_properties Any event properties to be processed.
+	 * @return stdClass The resulting event object with processed properties.
+	 */
+	protected static function process_properties(
+		string $event_name,
+		array $event_properties
+	): stdClass {
+		$event = (object) static::sanitize_properties_array( $event_properties );
+		$event = static::set_user_properties( $event );
+
+		// Set event name.
+		$event->_en = preg_replace(
+			'/^(?:' . static::EVENT_NAME_PREFIX . ')?(.*)/',
+			static::EVENT_NAME_PREFIX . '\1',
+			$event_name
+		) ?? '';
+
+		// Set event timestamp.
+		if ( ! isset( $event->_ts ) ) {
+			$event->_ts = static::get_timestamp();
+		}
+
+		// Remove non-routable IPs to prevent record from being discarded.
+		if ( property_exists( $event, '_via_ip' ) &&
+			1 === preg_match( '/^192\.168|^10\./', $event->_via_ip ) ) {
+			unset( $event->_via_ip );
+		}
+
+		// Set VIP environment if it exists.
+		if ( defined( 'VIP_GO_APP_ENVIRONMENT' ) ) {
+			$app_environment = constant( 'VIP_GO_APP_ENVIRONMENT' );
+			if ( is_string( $app_environment ) && '' !== $app_environment ) {
+				$event->vipgo_env = $app_environment;
+			}
+		}
+
```

Comment:
We should also add:

- org id
- if the user is a VIP user

Reply:
Added these in 2ebe3ad40e25314db8768e48ca88fd16a10109c0 but I'm not so sure about the VIP user check. I vaguely recall a helper function for this but the only thing I can find is `is_automattician()` which doesn't seem quite correct 🤔

Reply:
The `user_has_vip_support_role()` method is probably the best one, although I'm surprised we don't have a helper function for that.

https://github.com/Automattic/vip-go-mu-plugins/blob/44e6293555331be77646e30ee2df97c961fcd34c/vip-support/class-vip-support-user.php#L833

Maybe we can add a new `is_proxied_vip_support_user()` helper function that adds on to `is_proxied_automattician()`?

Reply:
Changed in e6a6436f60695cf401336128bc956eda060bea53